lib/internal/streams/web.js

```diff
@@ -0,0 +1,4433 @@
+      const preventCancel = options?.preventCancel;
+      const preventClose = options?.preventClose;
+      const signal = options?.signal;
+
+      if (signal !== undefined && signal?.[kAborted] === undefined)
+        throw new ERR_INVALID_ARG_TYPE('options.signal', 'AbortSignal', signal);
+
+      if (isReadableStreamLocked(this))
+        throw new ERR_INVALID_STATE.TypeError('The ReadableStream is locked');
+      if (isWritableStreamLocked(destination))
+        throw new ERR_INVALID_STATE.TypeError('The WritableStream is locked');
+
+      return readableStreamPipeTo(
+        this,
+        destination,
+        !!(preventClose),
+        !!(preventAbort),
+        !!(preventCancel),
+        signal);
+    } catch (error) {
+      return PromiseReject(error);
+    }
+  }
+
+  /**
+   * @returns {ReadableStream[]}
+   */
+  tee() {
+    if (!isReadableStream(this))
+      throw new ERR_INVALID_THIS('ReadableStream');
+    return readableStreamTee(this, false);
+  }
+
+  // The values() method is not defined in the streams standard
+  // spec but it is called out in the section on async iteration.
+  // So... that's weird.
+
+  /**
+   * @param {{
+   *   preventCancel? : boolean,
+   * }} [options]
+   * @returns {AsyncIterable}
+   */
+  values(options = {}) {
+    if (!isReadableStream(this))
+      throw new ERR_INVALID_THIS('ReadableStream');
+    validateObject(options, 'options');
+    const {
+      preventCancel = false,
+    } = options;
+
+    const reader = new ReadableStreamDefaultReader(this);
+
+    return ObjectSetPrototypeOf({
+      next() {
+        if (reader[kState].stream === undefined) {
+          return PromiseReject(
+            new ERR_INVALID_STATE.TypeError(
+              'The reader is not bound to a ReadableStream'));
+        }
+        const promise = createDeferredPromise();
+        readableStreamDefaultReaderRead(reader, {
+          [kChunk](chunk) {
+            promise.resolve({ value: chunk, done: false });
+          },
+          [kClose]() {
+            readableStreamReaderGenericRelease(reader);
+            promise.resolve({ done: true });
+          },
+          [kError](error) {
+            readableStreamReaderGenericRelease(reader);
+            promise.reject(error);
+          }
+        });
+        return promise.promise;
+      },
+
+      return(error) {
+        if (reader[kState].stream === undefined) {
+          return PromiseResolve();
+        }
+        assert(!reader[kState].readRequests.length);
+        if (!preventCancel) {
+          const result = readableStreamReaderGenericCancel(reader, error);
+          readableStreamReaderGenericRelease(reader);
+          return PromisePrototypeThen(
+            result,
+            () => { return { done: true }; });
+        }
+        readableStreamReaderGenericRelease(reader);
+        return PromiseResolve({ done: true });
+      },
+
+      [SymbolAsyncIterator]() { return this; }
+    }, AsyncIteratorPrototype);
+  }
```

Comment:
Rather than creating a custom iterator can't you just implement this as an async generator? Similar to what we do in node streams?

Reply:
That could work. The async generator machinery takes care of queuing multiple calls to `next()` or `return()` through `AsyncGeneratorEnqueue` ([example](https://tc39.es/ecma262/#sec-asyncgenerator-prototype-next)), so you'd get that part of the required Web IDL semantics for free. However, you'd have to be careful to match the timings of all the promise resolutions and side effects, e.g. ensuring that the reader's lock is released **before** you yield `{ done: true }` or throw an error.

Reply:
For now, the current implementation works. We can look at the possibility of switching to an async generator later.